pdl_syntax.md

```diff
@@ -0,0 +1,819 @@
+  doubles: array[double]
+  booleans: array[boolean]
+  strings: array[string]
+  bytes: array[bytes]
+}
+```
+
+Primitive arrays with default values: 
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+record WithPrimitivesArrayDefaults {
+  ints: array[int] = [1]
+  longs: array[long] = [3000000000]
+  floats: array[float] = [3.3]
+  doubles: array[double] = [4.4E38]
+  booleans: array[boolean] = [true]
+  strings: array[string] = ["hello"]
+  bytes: array[bytes] = ["\u0007"]
+}
+```
+
+Record arrays:
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+
+record WithRecordArray {
+  empties: array[Empty]
+  fruits: array[Fruits]
+}
+```
+
+Record arrays with default values: 
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Simple
+
+record WithRecordArrayDefaults {
+  empties: array[Simple] = [{ "message": "defaults!" }]
+  fruits: array[Fruits] = ["APPLE"]
+}
+```
+
+
+## Map Type
+
+Pegasus maps are defined with a values type and an optional key type.
+
+For example: 
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+record WithPrimitivesMap {
+  ints: map[string, int]
+  longs: map[string, long]
+  floats: map[string, float]
+  doubles: map[string, double]
+  booleans: map[string, boolean]
+  strings: map[string, string]
+  bytes: map[string, bytes]
+}
+```
+
+Primitive maps with default values: 
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+record WithPrimitivesMapDefaults {
+  ints: map[string, int] = { "int": 1 }
+  longs: map[string, long] = { "long": 3000000000 }
+  floats: map[string, float] = { "float": 3.3 }
+  doubles: map[string, double] = { "double": 4.4E38 }
+  booleans: map[string, boolean] = { "boolean": true }
+  strings: map[string, string] = { "string": "hello" }
+  bytes: map[string, bytes] = { "bytes": "\u0007" }
+}
+```
+
+Complex Types Map:
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+import com.linkedin.pegasus.generator.test.idl.records.Simple
+import com.linkedin.pegasus.generator.test.idl.`fixed`.Fixed8
+
+record WithComplexTypesMap {
+  empties: map[string, Empty]
+  fruits: map[string, Fruits]
+  arrays: map[string, array[Simple]]
+  maps: map[string, map[string, Simple]]
+  unions: map[
+    string,
+    typeref WithComplexTypesMapUnion = union[int, string, Simple]
+  ]
+  `fixed`: map[string, Fixed8]
+}
+```
+
+**Note**: 
+The key must always be "string".
+
+
+## Union Type
+
+A union type may be defined with any number of member types. Each member may be any pegasus type except union type. 
+Each member can be primitive, record, enum, map or array type.
```

Comment:
Suggested rewording:
A union type may be defined with any number of member types. Member type can be primitive, record, enum, map or array. Unions are not allowed as members inside an union.

Reply:
Updated.